at-angular.ts

```diff
@@ -107,8 +151,36 @@ module at {
             if (target.$inject && target.$inject.length > 0) {
                 factory.$inject = target.$inject.slice(0);
             }
-            angular.module(moduleName).factory(className, factory);
+            angular.module(moduleName).factory(className || getFuncName(target), factory);
+        };
+    }
+
+    export function decorator(moduleName: string, targetProvider: string): at.IClassAnnotationDecorator {
+
+        return (targetClass: any): void => {
+
+            angular
+                .module(moduleName)
+                .config([
+                    '$provide',
+                    function($provide: angular.auto.IProvideService): void {
+
+                    $provide.decorator(targetProvider, [
+                       '$delegate',
+                        ($delegate: any): void => {
+
+                            $delegate[targetProvider] = $delegate;
```

Comment:
Regarding your example, if `targetProvider` is `"SomeService"` line 172 will add reference to self on `SomeService.SomeService`. Why is it needed?